crates/ra_mbe/src/tt_cursor.rs

```diff
@@ -78,6 +86,28 @@ impl<'a> TtCursor<'a> {
         })
     }
 
+    pub(crate) fn eat_path(&mut self) -> Option<tt::Subtree> {
+        if self.subtree.token_trees.len() <= self.pos {
+            return None;
+        }
+
+        let subset = self.subtree.token_trees[self.pos..].to_vec();
+        let subtree = tt::Subtree { delimiter: tt::Delimiter::None, token_trees: subset };
+
+        let src = partial_parse(ra_syntax::SyntaxKind::PATH, &subtree);
+        let (sub_tree, _) = syntax_node_to_token_tree(&src)?;
+
+        for tt in sub_tree.token_trees.iter() {
+            if self.at_token(tt) {
+                self.bump()
+            } else {
+                return None;
+            }
+        }
```

Comment:
This logic feels a bit awkward... We clone the tokens, and than reverse-engenieer the number of consumed tokens. Idally, we should pass a cursor to the parser, and be able to tell, after the parser has finished, how much of data was consumed .

Reply:
In other words, the fact that TtTokenSource works by producing a vector of tokens is not a big design idea: it was just the simplest approach I came up with for the prototype :)

Reply:
While writing this PR, i just want to minimize changes.  I think i would implement a new `TokenSource` which is based on `tt::TokenTree` without convert it back and forth.